opacus/privacy_engine.py

```diff
@@ -31,7 +33,25 @@ def forbid_accumulation_hook(module: nn.Module, _):
 class PrivacyEngine:
     def __init__(self, secure_mode=False):
         self.accountant = RDPAccountant()
-        self.secure_mode = secure_mode  # TODO: actually support it
+        self.secure_mode = secure_mode
+
+        if self.secure_mode:
+            try:
+                import torchcsprng as csprng
+            except ImportError as e:
+                msg = (
+                    "To use secure RNG, you must install the torchcsprng package! "
+                    "Check out the instructions here: https://github.com/pytorch/csprng#installation"
+                )
+                raise ImportError(msg) from e
+
+            self.secure_rng = csprng.create_random_device_generator("/dev/urandom")
+        else:
+            warnings.warn(
```

Comment:
TIL. I always considered `warning` to be more of an internal program "problem" that can be fixed by user (such as deprecation, future proofing, etc.) whereas `logging.warn` as more of non-problem that is acceptable, but not desirable.
The documentation makes it clearer and I agree with you and Alex that warnings.warn sounds like the right way do it.

That said, since we explicitly allow users to not secure mode via a flag, it is best to document that we will throw a warning when secure mode is off. To a new user, it will be baffling if the freshly baked opacus with default settings throws warnings.

Reply:
> We should def. nudge users towards using secure_mode and this is a useful reminder

I'd argue that it highly depends on the use case. If you're doing research, you probably don't care and it's just annoying.
But since warnings are mostly treated the same as logged warnings, I agree it's ok to keep it as wanings. (and if your app is big enough that you want to do logger config, you should probably also consider using secure_mode)